074-connector-offsets-support.md

```diff
@@ -0,0 +1,180 @@
+* DELETE /connectors/{connector}/offsets
+* PUT /connectors/{connector}/stop
+
+Support for the "stop" endpoint was added via [PR 9095](https://github.com/strimzi/strimzi-kafka-operator/pull/9095).
+Currently users that are utilising the KafkaConnector resource cannot make use of the API endpoints for managing offsets.
+
+## Motivation
+
+Previously if users wanted to list, update or delete offsets for a connector they had to interact with the relevant Kafka topics directly.
+[KIP-875][kip] adds new endpoints to let the user do this more easily.
+There are multiple reasons a user might want to manage offsets, for example (as listed in the KIP):
+* Resetting the offsets for a connector while iterating rapidly in a development environment (so that the connector does not have to be renamed each time)
+* Viewing the in-memory offsets for source connectors on the cluster in order to recover from accidental deletion of the Connect source offsets topic (this is currently possible with the config topic, and can be a life-saver)
+* Monitoring the progress of running connectors on a per-source-partition basis 
+* Skipping records that cause issues with a connector and cannot be addressed using existing error-handling features
+
+## Proposal
+
+Strimzi should include mechanisms to let users manage offsets when utilising the KafkaConnector resources.
+This proposal suggests different approaches for each action, since they are all slightly different.
+In summary:
+ - listing offsets via the KafkaConnector status and enabled/disabled through a property
+ - altering offsets via a new resource called KafkaConnectorOffsets
+ - deleting offsets via an annotation on the KafkaConnector resource
+
+### Request/Response from the API endpoints
+
+The request/response used when interacting with the Kafka Connect API is different for source and sink connectors.
+This is because sink connectors use regular Kafka offsets, whereas source connectors determine their own structure for offsets.
+This difference is reflected in the approaches outlined in this proposal.
+
+Source connector format is:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        // Connector-defined source partition
+      },
+      "offset": {
+        // Connector-defined source offset
+      }
+    }
+  ]
+}
+```
+
+Sink connector format is:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "kafka_topic": // Kafka topic
+        "kafka_partition": // Kafka partition
+      },
+      "offset": {
+        "kafka_offset": // Kafka offset
+      }
+    }
+  ]
+}
+```
+
+### Listing offsets
+The KafkaConnector resource will be updated with a new property to enable the current offsets to be listed in the resource status.
+
+The new property will be a boolean property called `showOffsets` and default to `false`. It should default to `false` because enabling it requires the operator to do an additional API call every time it reconciles the connector.
```

Comment:
`every time it reconciles the connector`, How often is that?

Reply:
I believe every 2 minutes (similar to the cluster operator), and then if something changes in the KafkaConnector or KafkaConnect CR.

I realise that isn't very often if users are wanting live updates of the offsets. However it's enough to check whether a connector is progressing or not. And if the user is wanting to list offsets to see where a connector got stuck then they only need it to be updated once, since the connector is likely in a failed state.

Reply:
Is it possible to set up the watch in some way so that it doesn't trigger a reconciliation when the offset part of the status is updated? Or could we perhaps have some configurable timeout value and we only list offsets once the timeout is hit?

Reply:
No, the watch fires at every update. The onUpdate method of the watch gives you the old and new resource, so I guess you could do some filtering there and ignore the watch if only the offsets differ. Btu we do not do anything like that anywhere else, so not sure we want to go that way.

You could probably have some timestamp check and for example update the offsets only if they are over 1 minute old ... we don't do things like that anywhere, but technically it should not be a problem I guess.

That said, none of this would solve the size issues related to the resource status. So we should first figure out if that is a problem or not.